### ISSUE TITLE
JSTS sometimes takes too long

### ISSUE THREAD

**Issue by ericsoco** [stamen/usopendata#35]
Not sure what we can do to optimize, but I want to take a pass. It takes upwards of 30s for some diffs to render, e.g. http://openredist.org/#/open-redist/California-State-Senate/4

**Comment by waldoj** [comment]
I'm dubious that we can do anything about it—I think this is just an inevitable byproduct of extremely detailed GeoJSON files. I imagine there may be some Javascript optimization wizardry to be done, but that's pretty substantially out of scope. :)

**Comment by ericsoco** [comment]
Yeah, for the purposes of this site, it might make sense to downsample `geojson` files for display.
Maybe this? https://github.com/maxogden/simplify-geojson

**Comment by waldoj** [comment]
Ha! I'd be tickled to use one of Max's projects here. :) (He worked for US Open Data. Used to be on the board.)

**Comment by ericsoco** [comment]
I gave it a quick shot:
https://github.com/stamen/usopendata/commit/ec4b41b8cfc8623d05ade40a2c755c990d47702e
Didn't work out.

Some maps, like [this one](http://openredist.org/#/open-redist/Virginia-congressional-districts/11), end up with a really weird dashed line result:

![screen shot 2016-09-21 at 9 46 50 am](https://cloud.githubusercontent.com/assets/1127259/18720363/684f4614-7fe0-11e6-99c7-4aa2893b3dc3.png)

Others, like [this one](http://openredist.org/#/open-redist/California-State-Senate/4), just error out completely. In this case the complaint from JSTS is `found non-noded intersection between LINESTRING ( -121.316457 38.299853, -121.316931 38.30105 ) and LINESTRING ( -121.316899 38.302561, -121.31672443551557 38.300528359308295 ) [ (-121.31672443551557, 38.3005283593083, undefined) ]`. Not sure what it means, but it appears the geometry became malformed -- could be that the shape was closed originally but the simplification caused it to become open?

Feel free to bounce off Max if he's game. This is getting in over my head...

**Comment by almccon** [comment]
Those internal dashed lines look like they used to be borders between shapes? It is trying to dissolve the yellow part of Virginia into a single polygon? Or were those supposed to remain separate?

If you're trying to dissolve, it probably just needs to have a tiny buffer applied to prevent those slivers.

**Comment by ericsoco** [comment]
@almccon correct. [I union before I diff](https://github.com/stamen/usopendata/blob/master/js/components/DiffMap.jsx#L77). However, those lines should still be visible in the final product. (See #30.)

So, it's interesting that simplifying in this case half-fixes my problem of erasing data on union. But it's not a solution, it's a byproduct.

**Comment by ericsoco** [comment]
Revisiting this...seems like JSTS operations would be a good candidate to push off to a web worker, to avoid locking the UI while the diff runs.
